ccl/sqlproxyccl: add ACL for allowed CIDR ranges

jaylim-crl opened the pull request:
#### ccl/sqlproxyccl: add ACL for allowed CIDR ranges

Previously, the proxy only supported file based IP allowlist entries. This
commit adds a new ACL for allowed CIDR ranges, and will only apply if the
allowlist file isn't supplied to the proxy. Unlike the file-based ACL, this
makes use of the per-tenant metadata to store allowed CIDR ranges (i.e. in
the new AllowedCIDRRanges field). By default, if there are no CIDR ranges, it
is expected to block all incoming connections.

#### ccl/sqlproxyccl: rename PrivateEndpoints to AllowedPrivateEndpoints

This commit renames PrivateEndpoints to AllowedPrivateEndpoints in the tenant
proto message to be consistent with AllowedCIDRRanges. In theory, we could have
done AllowedEndpoints, but we'll leave "Private" in to be more explicit.

Release note: None

Epic: none

jaylim-crl committed:
ccl/sqlproxyccl: add ACL for allowed CIDR ranges

Previously, the proxy only supported file based IP allowlist entries. This
commit adds a new ACL for allowed CIDR ranges, and will only apply if the
allowlist file isn't supplied to the proxy. Unlike the file-based ACL, this
makes use of the per-tenant metadata to store allowed CIDR ranges (i.e. in
the new AllowedCIDRRanges field). By default, if there are no CIDR ranges, it
is expected to block all incoming connections.

Release note: None

Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103550)
<!-- Reviewable:end -->

jaylim-crl commented on pkg/ccl/sqlproxyccl/tenant/directory.proto:
```diff
@@ -129,10 +129,14 @@ message Tenant {
   // that could access the tenant. This will only apply if the tenant allows
   // private connections.
-  repeated string private_endpoints = 5;
+  repeated string private_endpoints = 6;
```
Nothing is consuming from this field yet, so we can just change it accordingly.

pjtatlow reviewed (changes requested):
Looking good! A few comments, happy to discuss any of them with you if you'd like.

jaylim-crl commented on pkg/ccl/sqlproxyccl/acl/public_cidr_ranges.go:
```diff
@@ -0,0 +1,66 @@
+	}
+
+	// By default, connections are rejected if the cluster does not allow public
+	// connections, or there are no ranges defined.
```
Done.

jaylim-crl commented on pkg/ccl/sqlproxyccl/acl/watcher.go:
```diff
@@ -186,13 +187,24 @@ func NewWatcher(ctx context.Context, opts ...Option) (*Watcher, error) {
+			cidrRangesController := &PublicCIDRRanges{
+				LookupTenantFn: w.options.lookupTenantFn,
+			}
+			w.addAccessController(ctx, cidrRangesController, pollAndUpdateChan(
```
Good point. Done. Added a comment as well.

jaylim-crl commented on pkg/ccl/sqlproxyccl/tenant/directory.proto:
```diff
@@ -129,10 +129,14 @@ message Tenant {
+  // PublicCIDRRanges corresponds to the list of source CIDR ranges that are
+  // allowed to access the tenant. This will only apply if the tenant allows
+  // public connections.
+  repeated string public_cidr_ranges = 5 [(gogoproto.customname) = "PublicCIDRRanges"];
```
I don't think we should link any of these concepts to what's in CC. "Public CIDR ranges" is what the proxy trying to enforce. In the context of Kubernetes, we use `LoadBalancerSourceRanges`. I'd like something that's generic here. How do you feel about `allowed_cidr_ranges`, `allowed_ranges`, or `source_ranges`? It would be nice to have the term `public` though, to distinguish itself from `private_endpoints`. If we had it has `ip_entries`, I would expect that to be actual IP addresses, and not CIDR ranges.

As for supporting SQL and UI, the same argument can be applied to private endpoints as well. For now, we will assume that both are for SQL only. Once we start supporting UI through the "SQL proxy" (which I think is unlikely [1]), we can revisit this decision. Note that these names are not encoded anywhere since they are just protobuf messages, so we have a lot of flexibility here.

[1] We would have more problems here since we're calling it a _SQL_ proxy.

pjtatlow replied:
I don't know that I necessarily want to link it to the same name just because it's in CC, but an allowlist is something people are familiar with. Like if someone saw the name "Public CIDR ranges", they would have no idea what that is. And if they asked you, you'd probably say "Oh, that's for the IP Allowlist". I really like the name `allowed_cidr_ranges` though, that's a lot more clear what it's there for!

[1] Haha yeah I think we might just call it the Serverless Proxy or something, but it seems like the most natural place to do it, since it already knows how to find tenants and route traffic to them. But that's a decision for another time.

jaylim-crl replied:
Done. I also renamed PrivateEndpoints to AllowedPrivateEndpoints.

jaylim-crl commented:
TFTR!

jaylim-crl committed:
ccl/sqlproxyccl: rename PrivateEndpoints to AllowedPrivateEndpoints

This commit renames PrivateEndpoints to AllowedPrivateEndpoints in the tenant
proto message to be consistent with AllowedCIDRRanges. In theory, we could have
done AllowedEndpoints, but we'll leave "Private" in to be more explicit.

Release note: None

Epic: none

pjtatlow reviewed (approved):
LGTM!

jaylim-crl commented:
TFTR!

bors r=pjtatlow

jaylim-crl commented:
bors r=pjtatlow
